src/transformers/models/grounding_dino/modeling_grounding_dino.py

```diff
@@ -0,0 +1,3768 @@
+
+    return attention_mask, position_ids.to(torch.long)
+
+
+@add_start_docstrings(
+    """
+    The bare Grounding DINO Model (consisting of a backbone and encoder-decoder Transformer) outputting raw
+    hidden-states without any specific head on top.
+    """,
+    GROUNDING_DINO_START_DOCSTRING,
+)
+class GroundingDINOModel(GroundingDINOPreTrainedModel):
+    def __init__(self, config: GroundingDINOConfig):
+        super().__init__(config)
+
+        # Create backbone + positional encoding
+        backbone = GroundingDINOConvEncoder(config)
+        position_embeddings = build_position_encoding(config)
+        self.backbone = GroundingDINOConvModel(backbone, position_embeddings)
+
+        # Create input projection layers
+        if config.num_feature_levels > 1:
+            num_backbone_outs = len(backbone.intermediate_channel_sizes)
+            input_proj_list = []
+            for _ in range(num_backbone_outs):
+                in_channels = backbone.intermediate_channel_sizes[_]
+                input_proj_list.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, config.d_model, kernel_size=1),
+                        nn.GroupNorm(32, config.d_model),
+                    )
+                )
+            for _ in range(config.num_feature_levels - num_backbone_outs):
+                input_proj_list.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, config.d_model, kernel_size=3, stride=2, padding=1),
+                        nn.GroupNorm(32, config.d_model),
+                    )
+                )
+                in_channels = config.d_model
+            self.input_proj_vision = nn.ModuleList(input_proj_list)
+        else:
+            self.input_proj_vision = nn.ModuleList(
+                [
+                    nn.Sequential(
+                        nn.Conv2d(backbone.intermediate_channel_sizes[-1], config.d_model, kernel_size=1),
+                        nn.GroupNorm(32, config.d_model),
+                    )
+                ]
+            )
+
+        # Create text backbone
+        self.text_backbone = GroundingDINOTextPrenet(config.text_backbone_config)
+        self.text_projection = nn.Linear(config.text_backbone_config.hidden_size, config.d_model)
+
+        if config.embedding_init_target or not config.two_stage:
+            self.query_position_embeddings = nn.Embedding(config.num_queries, config.d_model)
+
+        self.encoder = GroundingDINOEncoder(config)
+        self.decoder = GroundingDINODecoder(config)
+
+        self.level_embed = nn.Parameter(torch.Tensor(config.num_feature_levels, config.d_model))
+
+        print("Two stage:", config.two_stage)
+
+        if config.two_stage:
+            self.enc_output = nn.Linear(config.d_model, config.d_model)
+            self.enc_output_norm = nn.LayerNorm(config.d_model)
+            self.encoder_output_bbox_embed = None
+            self.encoder_output_class_embed = None
```

Comment:
Any reason this is set to None here? Cause in the forward pass, when `config.two_stage` is set to `True`, this class is called.

Reply:
Those are set in the `GroundingDINOForObjectDetection` when `config.two_stage` is set to `True` mainly because there is another parameter in the config called `two_stage_bbox_embed_share`, otherwise we could use the exact implementation from `DeformableDetr` that creates `config.decoder_layers + 1`, and to stick with the original implementation logic, since I found it easier to understand. What do you think?

Reply:
This makes some tests in tests/models/grounding_dino fail, so I guess we'll need to fix this. As said above, the class is called in forward hence it cannot be set to None

Reply:
Yeah, I've just fixed this and I'm running the tests again. I'll push the changes soon